compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,91 @@
+use rustc_abi::Abi;
+use rustc_middle::ty::{self, Instance, InstanceKind, ParamEnv, Ty, TyCtxt};
+use rustc_span::def_id::DefId;
+use rustc_span::{Span, Symbol};
+use rustc_target::abi::call::{FnAbi, PassMode};
+
+use crate::errors::{AbiErrorDisabledVectorTypeCall, AbiErrorDisabledVectorTypeDef};
+
+fn do_check_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    abi: &FnAbi<'tcx, Ty<'tcx>>,
+    target_feature_def: DefId,
+    emit_err: impl Fn(&'static str),
+) {
+    // FIXME: add support for other architectures
+    if tcx.sess.target.arch != "x86" && tcx.sess.target.arch != "x86_64" {
+        return;
+    }
+    let codegen_attrs = tcx.codegen_fn_attrs(target_feature_def);
+    for arg_abi in abi.args.iter().chain(std::iter::once(&abi.ret)) {
+        let size = arg_abi.layout.size;
+        if matches!(arg_abi.layout.abi, Abi::Vector { .. })
+            && matches!(arg_abi.mode, PassMode::Direct(_))
+        {
+            let feature = match size.bits() {
+                x if x <= 128 => "sse",
+                x if x <= 256 => "avx",
+                x if x <= 512 => "avx512f",
+                _ => {
+                    panic!("Unknown vector size for x86: {}; arg = {:?}", size.bits(), arg_abi)
+                }
```

Comment:
This should still be an array somewhere, not just code, to clearly separate the required data from the code that enforces the check.

To land the PR, we probably should cover at least all tier 1 architectures: x86, x86_64, and aarch64. Then we can have a tracking issue for filling in this information for other targets.

Reply:
I believe I did that.
In particular, to the best of my understanding (from reading the relevant aarch64 [procedure call standards](https://github.com/ARM-software/abi-aa/blob/2a70c42d62e9c3eb5887fa50b71257f20daca6f9/aapcs64/aapcs64.rst)), aarch64 ABI does *not* depend on features, so this check can be a noop there.

Reply:
Isn't there a feature for neon? Or does aarch64 just always have neon?

`-Ctarget-feature=-neon` might still work though and we should catch that...

Reply:
Indeed if you -Ctarget-feature=-neon you can run into issues. I modified the aarch64 path to check neon on aarch64.